napari/layers/_points_layer/model.py

```diff
@@ -426,6 +431,29 @@ def get_message(self, coord, value):
             msg = msg + ', index ' + str(value)
         return msg
 
+    def _update_thumbnail(self):
+        """Update thumbnail with current points and colors.
+        """
+        colormapped = np.zeros(self._thumbnail_shape)
+        colormapped[..., 3] = 1
+        if len(self._points_view) > 0:
+            min_vals = [self.range[-2][0], self.range[-1][0]]
+            shape = np.ceil(
+                [
+                    self.range[-2][1] - self.range[-2][0] + 1,
+                    self.range[-1][1] - self.range[-1][0] + 1,
+                ]
+            ).astype(int)
+            zoom_factor = np.divide(self._thumbnail_shape[:2], shape).min()
+            coords = np.floor(
+                (self._points_view - min_vals + 0.5) * zoom_factor
```

Comment:
Why use np.floor instead of np.round?

Reply:
I'm worried about the coords of a point getting rounded up and being equal to the shape of the thumbnail, which means I'll get an indexing error when I try and set that value

Reply:
This feels like a job for np.clip *and* a test. ;)

Reply:
I've added the `np.clip` to be sure, not sure what additional purpose the test serves!

Reply:
It means that when I come around 6 months from now and go "Hmm, this np.clip seems pointless, imma remove it to make the code prettier", I'll know immediately that I'm an idiot. =)

I think we're noticing with our bug whack-a-mole that we are at the stage where extensive testing is sorely needed — seemingly unrelated changes can end up breaking the code.

Reply:
agreed we should add more tests. once we've pulled out the vispy code from our layer model files we'll be able to more easily test our models and the functions inside them